{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "vep_cols = \"\"\"\\\n",
    "Allele|Consequence|IMPACT|SYMBOL|Gene|Feature_type|Feature|BIOTYPE|EXON\\\n",
    "|INTRON|HGVSc|HGVSp|cDNA_position|CDS_position|Protein_position|Amino_acids\\\n",
    "|Codons|Existing_variation|DISTANCE|STRAND|FLAGS|SYMBOL_SOURCE|HGNC_ID|CANONICAL\\\n",
    "|CCDS|HGVS_OFFSET\\\n",
    "\"\"\"\n",
    "vep_cols = vep_cols.split(\"|\")\n",
    "vep_cols = [term.strip().capitalize() for term in vep_cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Impact</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Biotype</th>\n",
       "      <th>Hgvsc</th>\n",
       "      <th>Hgvsp</th>\n",
       "      <th>Hgnc_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "      <td>117105737</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000546407</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7</td>\n",
       "      <td>117105737</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000546407</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>117105737</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000546407</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "      <td>117105737</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000546407</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7</td>\n",
       "      <td>117105737</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000546407</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1884</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #CHROM        POS REF ALT            Consequence    Impact Symbol  \\\n",
       "0       7  117105737   C   A  upstream_gene_variant  MODIFIER   CFTR   \n",
       "1       7  117105737   C   G  upstream_gene_variant  MODIFIER   CFTR   \n",
       "2       7  117105737   C   T  upstream_gene_variant  MODIFIER   CFTR   \n",
       "3       7  117105737   C  CA  upstream_gene_variant  MODIFIER   CFTR   \n",
       "4       7  117105737   C  CG  upstream_gene_variant  MODIFIER   CFTR   \n",
       "\n",
       "              Gene Feature_type          Feature               Biotype Hgvsc  \\\n",
       "0  ENSG00000001626   Transcript  ENST00000546407  processed_transcript         \n",
       "1  ENSG00000001626   Transcript  ENST00000546407  processed_transcript         \n",
       "2  ENSG00000001626   Transcript  ENST00000546407  processed_transcript         \n",
       "3  ENSG00000001626   Transcript  ENST00000546407  processed_transcript         \n",
       "4  ENSG00000001626   Transcript  ENST00000546407  processed_transcript         \n",
       "\n",
       "  Hgvsp Hgnc_id  \n",
       "0          1884  \n",
       "1          1884  \n",
       "2          1884  \n",
       "3          1884  \n",
       "4          1884  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vep_df = pd.read_table('./cftr.grch37.vep.vcf', header=0, skiprows=6, usecols=range(10))\n",
    "info_df = vep_df[\"INFO\"].str.replace(\"ANN=\", \"\").str.split(\",\").apply(pd.Series, 1).stack()\n",
    "info_df = info_df.str.split(\"|\").apply(pd.Series, 1)\n",
    "info_df.index = info_df.index.droplevel(-1)\n",
    "info_df.columns = vep_cols\n",
    "vep_df = vep_df.join(info_df)\n",
    "vep_df = vep_df[['#CHROM', 'POS', 'REF', 'ALT', 'Consequence', 'Impact', \n",
    "                     'Symbol', 'Gene', 'Feature_type', 'Feature', 'Biotype', \n",
    "                     'Hgvsc','Hgvsp','Hgnc_id']]\n",
    "vep_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "snpeff_cols = \"\"\"\\\n",
    "Allele|Annotation|Putative_impact|Gene_name|Gene_ID|Feature_type|Feature_ID|Transcript_Biotype|Rank|HGVSc|HGVSp\\\n",
    "|cDNA_position|CDS_position|Protein_position|Distance|Errors\"\"\"\n",
    "snpeff_cols = snpeff_cols.split(\"|\")\n",
    "snpeff_cols = [term.strip().capitalize() for term in snpeff_cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Putative_impact</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Feature_id</th>\n",
       "      <th>Transcript_biotype</th>\n",
       "      <th>Hgvsc</th>\n",
       "      <th>Hgvsp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "      <td>117105737</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000546407</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>n.-101C&gt;A</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "      <td>117105737</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>intergenic_region</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>ENSG00000214684-ENSG00000001626</td>\n",
       "      <td>intergenic_region</td>\n",
       "      <td>ENSG00000214684-ENSG00000001626</td>\n",
       "      <td></td>\n",
       "      <td>n.117105737C&gt;A</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7</td>\n",
       "      <td>117105737</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000546407</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>n.-101C&gt;G</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7</td>\n",
       "      <td>117105737</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>intergenic_region</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>ENSG00000214684-ENSG00000001626</td>\n",
       "      <td>intergenic_region</td>\n",
       "      <td>ENSG00000214684-ENSG00000001626</td>\n",
       "      <td></td>\n",
       "      <td>n.117105737C&gt;G</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>117105737</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000546407</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>n.-101C&gt;T</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #CHROM        POS REF ALT             Annotation Putative_impact  \\\n",
       "0       7  117105737   C   A  upstream_gene_variant        MODIFIER   \n",
       "0       7  117105737   C   A      intergenic_region        MODIFIER   \n",
       "1       7  117105737   C   G  upstream_gene_variant        MODIFIER   \n",
       "1       7  117105737   C   G      intergenic_region        MODIFIER   \n",
       "2       7  117105737   C   T  upstream_gene_variant        MODIFIER   \n",
       "\n",
       "                           Gene_id       Feature_type  \\\n",
       "0                  ENSG00000001626         transcript   \n",
       "0  ENSG00000214684-ENSG00000001626  intergenic_region   \n",
       "1                  ENSG00000001626         transcript   \n",
       "1  ENSG00000214684-ENSG00000001626  intergenic_region   \n",
       "2                  ENSG00000001626         transcript   \n",
       "\n",
       "                        Feature_id    Transcript_biotype           Hgvsc Hgvsp  \n",
       "0                  ENST00000546407  processed_transcript       n.-101C>A        \n",
       "0  ENSG00000214684-ENSG00000001626                        n.117105737C>A        \n",
       "1                  ENST00000546407  processed_transcript       n.-101C>G        \n",
       "1  ENSG00000214684-ENSG00000001626                        n.117105737C>G        \n",
       "2                  ENST00000546407  processed_transcript       n.-101C>T        "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snpeff_df = pd.read_table('./cftr.grch37.snpeff.vcf', header=0, skiprows=9, usecols=range(10))\n",
    "info_df = snpeff_df[\"INFO\"].str.split(\";\").apply(pd.Series, 1)[0] #snpeff includes two other INFO fields that we don't need\n",
    "info_df = info_df.str.replace(\"ANN=\", \"\").str.split(\",\").apply(pd.Series, 1).stack()\n",
    "info_df = info_df.str.split(\"|\").apply(pd.Series, 1)\n",
    "info_df.index = info_df.index.droplevel(-1)\n",
    "info_df.columns = snpeff_cols\n",
    "snpeff_df = snpeff_df.join(info_df)\n",
    "snpeff_df = snpeff_df[['#CHROM', 'POS', 'REF', 'ALT', 'Annotation', 'Putative_impact','Gene_id', 'Feature_type', 'Feature_id', 'Transcript_biotype', \n",
    "                     'Hgvsc','Hgvsp']]\n",
    "snpeff_df.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
