{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "vep_cols = \"\"\"\\\n",
    "Allele|Consequence|IMPACT|SYMBOL|Gene|Feature_type|Feature|BIOTYPE|EXON\\\n",
    "|INTRON|HGVSc|HGVSp|cDNA_position|CDS_position|Protein_position|Amino_acids\\\n",
    "|Codons|Existing_variation|DISTANCE|STRAND|FLAGS|SYMBOL_SOURCE|HGNC_ID|CANONICAL\\\n",
    "|CCDS|HGVS_OFFSET\\\n",
    "\"\"\"\n",
    "vep_cols = vep_cols.split(\"|\")\n",
    "vep_cols = [term.strip().capitalize() for term in vep_cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Impact</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Biotype</th>\n",
       "      <th>Hgvsc</th>\n",
       "      <th>Hgvsp</th>\n",
       "      <th>Hgnc_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "      <td>117105737</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000546407</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7</td>\n",
       "      <td>117105737</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000546407</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>117105737</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000546407</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "      <td>117105737</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000546407</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7</td>\n",
       "      <td>117105737</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000546407</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1884</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #CHROM        POS REF ALT            Consequence    Impact Symbol  \\\n",
       "0       7  117105737   C   A  upstream_gene_variant  MODIFIER   CFTR   \n",
       "1       7  117105737   C   G  upstream_gene_variant  MODIFIER   CFTR   \n",
       "2       7  117105737   C   T  upstream_gene_variant  MODIFIER   CFTR   \n",
       "3       7  117105737   C  CA  upstream_gene_variant  MODIFIER   CFTR   \n",
       "4       7  117105737   C  CG  upstream_gene_variant  MODIFIER   CFTR   \n",
       "\n",
       "              Gene Feature_type          Feature               Biotype Hgvsc  \\\n",
       "0  ENSG00000001626   Transcript  ENST00000546407  processed_transcript         \n",
       "1  ENSG00000001626   Transcript  ENST00000546407  processed_transcript         \n",
       "2  ENSG00000001626   Transcript  ENST00000546407  processed_transcript         \n",
       "3  ENSG00000001626   Transcript  ENST00000546407  processed_transcript         \n",
       "4  ENSG00000001626   Transcript  ENST00000546407  processed_transcript         \n",
       "\n",
       "  Hgvsp Hgnc_id  \n",
       "0          1884  \n",
       "1          1884  \n",
       "2          1884  \n",
       "3          1884  \n",
       "4          1884  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vep_df = pd.read_table('./cftr.grch37.vep.vcf', header=0, skiprows=6, usecols=range(10))\n",
    "info_df = vep_df[\"INFO\"].str.replace(\"ANN=\", \"\").str.split(\",\").apply(pd.Series, 1).stack()\n",
    "info_df = info_df.str.split(\"|\").apply(pd.Series, 1)\n",
    "info_df.index = info_df.index.droplevel(-1)\n",
    "info_df.columns = vep_cols\n",
    "vep_df = vep_df.join(info_df)\n",
    "vep_df = vep_df[['#CHROM', 'POS', 'REF', 'ALT', 'Consequence', 'Impact', \n",
    "                     'Symbol', 'Gene', 'Feature_type', 'Feature', 'Biotype', \n",
    "                     'Hgvsc','Hgvsp','Hgnc_id']]\n",
    "del info_df\n",
    "vep_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "snpeff_cols = \"\"\"\\\n",
    "Allele|Annotation|Putative_impact|Gene_name|Gene_ID|Feature_type|Feature_ID|Transcript_Biotype|Rank|HGVSc|HGVSp\\\n",
    "|cDNA_position|CDS_position|Protein_position|Distance|Errors\"\"\"\n",
    "snpeff_cols = snpeff_cols.split(\"|\")\n",
    "snpeff_cols = [term.strip().capitalize() for term in snpeff_cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Putative_impact</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Feature_id</th>\n",
       "      <th>Transcript_biotype</th>\n",
       "      <th>Hgvsc</th>\n",
       "      <th>Hgvsp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "      <td>117105737</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000546407</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>n.-101C&gt;A</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "      <td>117105737</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>intergenic_region</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>ENSG00000214684-ENSG00000001626</td>\n",
       "      <td>intergenic_region</td>\n",
       "      <td>ENSG00000214684-ENSG00000001626</td>\n",
       "      <td></td>\n",
       "      <td>n.117105737C&gt;A</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7</td>\n",
       "      <td>117105737</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000546407</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>n.-101C&gt;G</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7</td>\n",
       "      <td>117105737</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>intergenic_region</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>ENSG00000214684-ENSG00000001626</td>\n",
       "      <td>intergenic_region</td>\n",
       "      <td>ENSG00000214684-ENSG00000001626</td>\n",
       "      <td></td>\n",
       "      <td>n.117105737C&gt;G</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>117105737</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000546407</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>n.-101C&gt;T</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #CHROM        POS REF ALT             Annotation Putative_impact  \\\n",
       "0       7  117105737   C   A  upstream_gene_variant        MODIFIER   \n",
       "0       7  117105737   C   A      intergenic_region        MODIFIER   \n",
       "1       7  117105737   C   G  upstream_gene_variant        MODIFIER   \n",
       "1       7  117105737   C   G      intergenic_region        MODIFIER   \n",
       "2       7  117105737   C   T  upstream_gene_variant        MODIFIER   \n",
       "\n",
       "                           Gene_id       Feature_type  \\\n",
       "0                  ENSG00000001626         transcript   \n",
       "0  ENSG00000214684-ENSG00000001626  intergenic_region   \n",
       "1                  ENSG00000001626         transcript   \n",
       "1  ENSG00000214684-ENSG00000001626  intergenic_region   \n",
       "2                  ENSG00000001626         transcript   \n",
       "\n",
       "                        Feature_id    Transcript_biotype           Hgvsc Hgvsp  \n",
       "0                  ENST00000546407  processed_transcript       n.-101C>A        \n",
       "0  ENSG00000214684-ENSG00000001626                        n.117105737C>A        \n",
       "1                  ENST00000546407  processed_transcript       n.-101C>G        \n",
       "1  ENSG00000214684-ENSG00000001626                        n.117105737C>G        \n",
       "2                  ENST00000546407  processed_transcript       n.-101C>T        "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snpeff_df = pd.read_table('./cftr.grch37.snpeff.vcf', header=0, skiprows=9, usecols=range(10))\n",
    "info_df = snpeff_df[\"INFO\"].str.split(\";\").apply(pd.Series, 1)[0] #snpeff includes two other INFO fields that we don't need\n",
    "info_df = info_df.str.replace(\"ANN=\", \"\").str.split(\",\").apply(pd.Series, 1).stack()\n",
    "info_df = info_df.str.split(\"|\").apply(pd.Series, 1)\n",
    "info_df.index = info_df.index.droplevel(-1)\n",
    "info_df.columns = snpeff_cols\n",
    "snpeff_df = snpeff_df.join(info_df)\n",
    "snpeff_df = snpeff_df[['#CHROM', 'POS', 'REF', 'ALT', 'Annotation', 'Putative_impact','Gene_id', 'Feature_type', 'Feature_id', 'Transcript_biotype', \n",
    "                     'Hgvsc','Hgvsp']]\n",
    "del info_df\n",
    "snpeff_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#From http://snpeff.sourceforge.net/VCFannotationformat_v1.0.pdf with additions:\n",
    "#VEP: non_coding_transcript_exon_variant, non_coding_transcript_variant, protein altering variant, \n",
    "#incomplete_terminal_codon_variant, NMD_transcript_variant\n",
    "#Snpeff: conservative_inframe_deletion, conservative_inframe_insertion, structural_interaction_variant, 5_prime_UTR_truncation\n",
    "\n",
    "ranked_terms = [\"chromosome_number_variation\",\"exon_loss_variant\",\"frameshift_variant\",\"stop_gained\",\"stop_lost\",\n",
    "                \"start_lost\",\"splice_acceptor_variant\",\"splice_donor_variant\",\"rare_amino_acid_variant\",\"missense_variant\",\n",
    "                \"inframe_insertion\",\"conservative_inframe_insertion\", \"disruptive_inframe_insertion\",\"inframe_deletion\",\"conservative_inframe_deletion\", \"disruptive_inframe_deletion\",\n",
    "                \"5_prime_UTR_truncation+exon_loss_variant\",\"5_prime_UTR_truncation\",\"3_prime_UTR_truncation+exon_loss\",\"splice_branch_variant\",\n",
    "                \"splice_region_variant\",\"splice_branch_variant\",\"stop_retained_variant\",\"initiator_codon_variant\",\n",
    "                \"synonymous_variant\",\"initiator_codon_variant+non_canonical_start_codon\",\"stop_retained_variant\",\n",
    "                \"coding_sequence_variant\",\"5_prime_UTR_variant\",\"3_prime_UTR_variant\",\n",
    "                \"5_prime_UTR_premature_start_codon_gain_variant\",\"structural_interaction_variant\", \"protein_altering_variant\",\"upstream_gene_variant\",\"downstream_gene_variant\",\n",
    "                \"TF_binding_site_variant\",\"regulatory_region_variant\",\"miRNA\",\"custom\",\"sequence_feature\",\n",
    "                \"conserved_intron_variant\",\"intron_variant\",\"intragenic_variant\",\"conserved_intergenic_variant\",\n",
    "                \"intergenic_region\",\"coding_sequence_variant\",\"non_coding_exon_variant\",\"non_coding_transcript_exon_variant\",\n",
    "                \"nc_transcript_variant\",\"non_coding_transcript_variant\",\"NMD_transcript_variant\", \"incomplete_terminal_codon_variant\", \"gene_variant\",\"chromosome\"]\n",
    "def term_rank(term):\n",
    "    return ranked_terms.index(term)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "vep_df[\"effect\"] = vep_df.apply(lambda row: min(row[\"Consequence\"].split(\"&\"), key=term_rank), axis=1)\n",
    "\n",
    "snpeff_df[\"effect\"] = snpeff_df.apply(lambda row: min(row[\"Annotation\"].split('&'), key=term_rank), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Annotations\n",
      "\n",
      "                                                     VEP    SnpEff\n",
      "3_prime_UTR_variant                               6636.0    6680.0\n",
      "5_prime_UTR_premature_start_codon_gain_variant       0.0     286.0\n",
      "5_prime_UTR_variant                               9969.0   10259.0\n",
      "coding_sequence_variant                             64.0       0.0\n",
      "conservative_inframe_deletion                        0.0    4619.0\n",
      "conservative_inframe_insertion                       0.0    8650.0\n",
      "disruptive_inframe_deletion                          0.0    8990.0\n",
      "disruptive_inframe_insertion                         0.0   17211.0\n",
      "downstream_gene_variant                          99177.0   99185.0\n",
      "exon_loss_variant                                    0.0       7.0\n",
      "frameshift_variant                              113706.0  113693.0\n",
      "incomplete_terminal_codon_variant                    8.0       0.0\n",
      "inframe_deletion                                 13604.0       0.0\n",
      "inframe_insertion                                18961.0       0.0\n",
      "initiator_codon_variant                              0.0       8.0\n",
      "intergenic_region                                    0.0    1403.0\n",
      "intron_variant                                  417918.0  416946.0\n",
      "missense_variant                                 30988.0   30988.0\n",
      "non_coding_transcript_exon_variant               29794.0   29567.0\n",
      "non_coding_transcript_variant                        0.0      42.0\n",
      "protein_altering_variant                          6834.0       0.0\n",
      "sequence_feature                                     0.0    5220.0\n",
      "splice_acceptor_variant                           2208.0    2776.0\n",
      "splice_donor_variant                              2208.0    2869.0\n",
      "splice_region_variant                            17865.0   18005.0\n",
      "start_lost                                          55.0      65.0\n",
      "stop_gained                                       4877.0    4867.0\n",
      "stop_lost                                           55.0      63.0\n",
      "stop_retained_variant                                5.0       5.0\n",
      "structural_interaction_variant                       0.0   78159.0\n",
      "synonymous_variant                                9056.0    9052.0\n",
      "upstream_gene_variant                            83853.0   83809.0\n",
      "\n",
      "Impacts\n",
      "             VEP  SnpEff\n",
      "HIGH      123109  202492\n",
      "LOW        26698   31676\n",
      "MODERATE   70645   71365\n",
      "MODIFIER  647389  647891\n",
      "\n",
      "Counts\n",
      "                         VEP    SnpEff\n",
      "#CHROM              867841.0  953424.0\n",
      "ALT                 867841.0  953424.0\n",
      "Annotation               NaN  953424.0\n",
      "Biotype             867841.0       NaN\n",
      "Consequence         867841.0       NaN\n",
      "Feature             867841.0       NaN\n",
      "Feature_id               NaN  953424.0\n",
      "Feature_type        867841.0  953424.0\n",
      "Gene                867841.0       NaN\n",
      "Gene_id                  NaN  953424.0\n",
      "Hgnc_id             867841.0       NaN\n",
      "Hgvsc               867841.0  953424.0\n",
      "Hgvsp               867841.0  953424.0\n",
      "Impact              867841.0       NaN\n",
      "POS                 867841.0  953424.0\n",
      "Putative_impact          NaN  953424.0\n",
      "REF                 867841.0  953424.0\n",
      "Symbol              867841.0       NaN\n",
      "Transcript_biotype       NaN  953424.0\n",
      "effect              867841.0  953424.0\n"
     ]
    }
   ],
   "source": [
    "vc_vep = vep_df.groupby(['effect']).size()\n",
    "vc_vep.name = \"VEP\"\n",
    "vc_snpeff = snpeff_df.groupby(['effect']).size()\n",
    "vc_snpeff.name = \"SnpEff\"\n",
    "vc_df = pd.DataFrame([vc_vep, vc_snpeff])\n",
    "print(\"Annotations\\n\")\n",
    "print(vc_df.transpose().fillna(0))\n",
    "impact_vep = vep_df.groupby(['Impact']).size()\n",
    "impact_vep.name = \"VEP\"\n",
    "impact_snpeff = snpeff_df.groupby(['Putative_impact']).size()\n",
    "impact_snpeff.name = \"SnpEff\"\n",
    "impact_df = pd.DataFrame([impact_vep, impact_snpeff])\n",
    "print(\"\\nImpacts\")\n",
    "print(impact_df.transpose())\n",
    "counts_vep = vep_df.count()\n",
    "counts_vep.name = 'VEP'\n",
    "counts_snpeff = snpeff_df.count()\n",
    "counts_snpeff.name = 'SnpEff'\n",
    "counts_df = pd.DataFrame([counts_vep, counts_snpeff])\n",
    "print(\"\\nCounts\")\n",
    "print(counts_df.transpose())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This example is interesting. Vep is usually a pretty general `protein_altering_variant` annotation. Snpeff breaks it down to `disruptive_inframe_insertion`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Impact</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Biotype</th>\n",
       "      <th>Hgvsc</th>\n",
       "      <th>Hgvsp</th>\n",
       "      <th>Hgnc_id</th>\n",
       "      <th>effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19511</th>\n",
       "      <td>7</td>\n",
       "      <td>117120152</td>\n",
       "      <td>C</td>\n",
       "      <td>CCGA</td>\n",
       "      <td>protein_altering_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000003084</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENST00000003084.6:c.4_5insCGA</td>\n",
       "      <td>ENSP00000003084.6:p.Gln2delinsProLys</td>\n",
       "      <td>1884</td>\n",
       "      <td>protein_altering_variant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       #CHROM        POS REF   ALT               Consequence    Impact Symbol  \\\n",
       "19511       7  117120152   C  CCGA  protein_altering_variant  MODERATE   CFTR   \n",
       "\n",
       "                  Gene Feature_type          Feature         Biotype  \\\n",
       "19511  ENSG00000001626   Transcript  ENST00000003084  protein_coding   \n",
       "\n",
       "                               Hgvsc                                 Hgvsp  \\\n",
       "19511  ENST00000003084.6:c.4_5insCGA  ENSP00000003084.6:p.Gln2delinsProLys   \n",
       "\n",
       "      Hgnc_id                    effect  \n",
       "19511    1884  protein_altering_variant  "
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vep_df[vep_df['effect'].str.contains('protein_altering_variant')][:1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Impact</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Biotype</th>\n",
       "      <th>Hgvsc</th>\n",
       "      <th>Hgvsp</th>\n",
       "      <th>Hgnc_id</th>\n",
       "      <th>effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19511</th>\n",
       "      <td>7</td>\n",
       "      <td>117120152</td>\n",
       "      <td>C</td>\n",
       "      <td>CCGA</td>\n",
       "      <td>protein_altering_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000003084</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENST00000003084.6:c.4_5insCGA</td>\n",
       "      <td>ENSP00000003084.6:p.Gln2delinsProLys</td>\n",
       "      <td>1884</td>\n",
       "      <td>protein_altering_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19511</th>\n",
       "      <td>7</td>\n",
       "      <td>117120152</td>\n",
       "      <td>C</td>\n",
       "      <td>CCGA</td>\n",
       "      <td>protein_altering_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000426809</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENST00000426809.1:c.4_5insCGA</td>\n",
       "      <td>ENSP00000389119.1:p.Gln2delinsProLys</td>\n",
       "      <td>1884</td>\n",
       "      <td>protein_altering_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19511</th>\n",
       "      <td>7</td>\n",
       "      <td>117120152</td>\n",
       "      <td>C</td>\n",
       "      <td>CCGA</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000446805</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENST00000446805.1:c.-191+404_-191+405insCGA</td>\n",
       "      <td></td>\n",
       "      <td>1884</td>\n",
       "      <td>intron_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19511</th>\n",
       "      <td>7</td>\n",
       "      <td>117120152</td>\n",
       "      <td>C</td>\n",
       "      <td>CCGA</td>\n",
       "      <td>protein_altering_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000454343</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENST00000454343.1:c.4_5insCGA</td>\n",
       "      <td>ENSP00000403677.1:p.Gln2delinsProLys</td>\n",
       "      <td>1884</td>\n",
       "      <td>protein_altering_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19511</th>\n",
       "      <td>7</td>\n",
       "      <td>117120152</td>\n",
       "      <td>C</td>\n",
       "      <td>CCGA</td>\n",
       "      <td>intron_variant&amp;non_coding_transcript_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000546407</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>ENST00000546407.1:n.166+4290_166+4291insCGA</td>\n",
       "      <td></td>\n",
       "      <td>1884</td>\n",
       "      <td>intron_variant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       #CHROM        POS REF   ALT  \\\n",
       "19511       7  117120152   C  CCGA   \n",
       "19511       7  117120152   C  CCGA   \n",
       "19511       7  117120152   C  CCGA   \n",
       "19511       7  117120152   C  CCGA   \n",
       "19511       7  117120152   C  CCGA   \n",
       "\n",
       "                                        Consequence    Impact Symbol  \\\n",
       "19511                      protein_altering_variant  MODERATE   CFTR   \n",
       "19511                      protein_altering_variant  MODERATE   CFTR   \n",
       "19511                                intron_variant  MODIFIER   CFTR   \n",
       "19511                      protein_altering_variant  MODERATE   CFTR   \n",
       "19511  intron_variant&non_coding_transcript_variant  MODIFIER   CFTR   \n",
       "\n",
       "                  Gene Feature_type          Feature               Biotype  \\\n",
       "19511  ENSG00000001626   Transcript  ENST00000003084        protein_coding   \n",
       "19511  ENSG00000001626   Transcript  ENST00000426809        protein_coding   \n",
       "19511  ENSG00000001626   Transcript  ENST00000446805        protein_coding   \n",
       "19511  ENSG00000001626   Transcript  ENST00000454343        protein_coding   \n",
       "19511  ENSG00000001626   Transcript  ENST00000546407  processed_transcript   \n",
       "\n",
       "                                             Hgvsc  \\\n",
       "19511                ENST00000003084.6:c.4_5insCGA   \n",
       "19511                ENST00000426809.1:c.4_5insCGA   \n",
       "19511  ENST00000446805.1:c.-191+404_-191+405insCGA   \n",
       "19511                ENST00000454343.1:c.4_5insCGA   \n",
       "19511  ENST00000546407.1:n.166+4290_166+4291insCGA   \n",
       "\n",
       "                                      Hgvsp Hgnc_id                    effect  \n",
       "19511  ENSP00000003084.6:p.Gln2delinsProLys    1884  protein_altering_variant  \n",
       "19511  ENSP00000389119.1:p.Gln2delinsProLys    1884  protein_altering_variant  \n",
       "19511                                          1884            intron_variant  \n",
       "19511  ENSP00000403677.1:p.Gln2delinsProLys    1884  protein_altering_variant  \n",
       "19511                                          1884            intron_variant  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vep_df.ix[19511]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Putative_impact</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Feature_id</th>\n",
       "      <th>Transcript_biotype</th>\n",
       "      <th>Hgvsc</th>\n",
       "      <th>Hgvsp</th>\n",
       "      <th>effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23333</th>\n",
       "      <td>7</td>\n",
       "      <td>117144330</td>\n",
       "      <td>A</td>\n",
       "      <td>ATCG</td>\n",
       "      <td>disruptive_inframe_insertion</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000003084</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>c.77_78insTCG</td>\n",
       "      <td>p.Lys26delinsAsnArg</td>\n",
       "      <td>disruptive_inframe_insertion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23333</th>\n",
       "      <td>7</td>\n",
       "      <td>117144330</td>\n",
       "      <td>A</td>\n",
       "      <td>ATCG</td>\n",
       "      <td>disruptive_inframe_insertion</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000454343</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>c.77_78insTCG</td>\n",
       "      <td>p.Lys26delinsAsnArg</td>\n",
       "      <td>disruptive_inframe_insertion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23333</th>\n",
       "      <td>7</td>\n",
       "      <td>117144330</td>\n",
       "      <td>A</td>\n",
       "      <td>ATCG</td>\n",
       "      <td>disruptive_inframe_insertion</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000426809</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>c.77_78insTCG</td>\n",
       "      <td>p.Lys26delinsAsnArg</td>\n",
       "      <td>disruptive_inframe_insertion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23333</th>\n",
       "      <td>7</td>\n",
       "      <td>117144330</td>\n",
       "      <td>A</td>\n",
       "      <td>ATCG</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000446805</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>c.-167_-166insTCG</td>\n",
       "      <td></td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23333</th>\n",
       "      <td>7</td>\n",
       "      <td>117144330</td>\n",
       "      <td>A</td>\n",
       "      <td>ATCG</td>\n",
       "      <td>non_coding_transcript_exon_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000546407</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>n.190_191insTCG</td>\n",
       "      <td></td>\n",
       "      <td>non_coding_transcript_exon_variant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       #CHROM        POS REF   ALT                          Annotation  \\\n",
       "23333       7  117144330   A  ATCG        disruptive_inframe_insertion   \n",
       "23333       7  117144330   A  ATCG        disruptive_inframe_insertion   \n",
       "23333       7  117144330   A  ATCG        disruptive_inframe_insertion   \n",
       "23333       7  117144330   A  ATCG                 5_prime_UTR_variant   \n",
       "23333       7  117144330   A  ATCG  non_coding_transcript_exon_variant   \n",
       "\n",
       "      Putative_impact          Gene_id Feature_type       Feature_id  \\\n",
       "23333        MODERATE  ENSG00000001626   transcript  ENST00000003084   \n",
       "23333        MODERATE  ENSG00000001626   transcript  ENST00000454343   \n",
       "23333        MODERATE  ENSG00000001626   transcript  ENST00000426809   \n",
       "23333        MODIFIER  ENSG00000001626   transcript  ENST00000446805   \n",
       "23333        MODIFIER  ENSG00000001626   transcript  ENST00000546407   \n",
       "\n",
       "         Transcript_biotype              Hgvsc                Hgvsp  \\\n",
       "23333        protein_coding      c.77_78insTCG  p.Lys26delinsAsnArg   \n",
       "23333        protein_coding      c.77_78insTCG  p.Lys26delinsAsnArg   \n",
       "23333        protein_coding      c.77_78insTCG  p.Lys26delinsAsnArg   \n",
       "23333        protein_coding  c.-167_-166insTCG                        \n",
       "23333  processed_transcript    n.190_191insTCG                        \n",
       "\n",
       "                                   effect  \n",
       "23333        disruptive_inframe_insertion  \n",
       "23333        disruptive_inframe_insertion  \n",
       "23333        disruptive_inframe_insertion  \n",
       "23333                 5_prime_UTR_variant  \n",
       "23333  non_coding_transcript_exon_variant  "
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snpeff_df.ix[19511]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So this example is interesting. It shows two things:\n",
    "1. SnpEff is providing these \"structural_interaction_variant\" annotations, which vep does not provide\n",
    "2. SnpEff is a bit more granular annotating inframe variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Putative_impact</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Feature_id</th>\n",
       "      <th>Transcript_biotype</th>\n",
       "      <th>Hgvsc</th>\n",
       "      <th>Hgvsp</th>\n",
       "      <th>effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>58085</th>\n",
       "      <td>7</td>\n",
       "      <td>117182112</td>\n",
       "      <td>TTAA</td>\n",
       "      <td>T</td>\n",
       "      <td>structural_interaction_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>interaction</td>\n",
       "      <td>2PZG:B_388-B_567:ENST00000003084</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>c.1160_1162delTAA</td>\n",
       "      <td></td>\n",
       "      <td>structural_interaction_variant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       #CHROM        POS   REF ALT                      Annotation  \\\n",
       "58085       7  117182112  TTAA   T  structural_interaction_variant   \n",
       "\n",
       "      Putative_impact          Gene_id Feature_type  \\\n",
       "58085            HIGH  ENSG00000001626  interaction   \n",
       "\n",
       "                             Feature_id Transcript_biotype              Hgvsc  \\\n",
       "58085  2PZG:B_388-B_567:ENST00000003084     protein_coding  c.1160_1162delTAA   \n",
       "\n",
       "      Hgvsp                          effect  \n",
       "58085        structural_interaction_variant  "
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snpeff_df[snpeff_df['effect'].str.contains('structural_interaction_variant')][:1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Impact</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Biotype</th>\n",
       "      <th>Hgvsc</th>\n",
       "      <th>Hgvsp</th>\n",
       "      <th>Hgnc_id</th>\n",
       "      <th>effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>58085</th>\n",
       "      <td>7</td>\n",
       "      <td>117182112</td>\n",
       "      <td>TTAA</td>\n",
       "      <td>T</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000003084</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENST00000003084.6:c.1160_1162delTAA</td>\n",
       "      <td>ENSP00000003084.6:p.Leu387_Thr388delinsSer</td>\n",
       "      <td>1884</td>\n",
       "      <td>inframe_deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58085</th>\n",
       "      <td>7</td>\n",
       "      <td>117182112</td>\n",
       "      <td>TTAA</td>\n",
       "      <td>T</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000426809</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENST00000426809.1:c.1070_1072delTAA</td>\n",
       "      <td>ENSP00000389119.1:p.Leu357_Thr358delinsSer</td>\n",
       "      <td>1884</td>\n",
       "      <td>inframe_deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58085</th>\n",
       "      <td>7</td>\n",
       "      <td>117182112</td>\n",
       "      <td>TTAA</td>\n",
       "      <td>T</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000454343</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENST00000454343.1:c.1160_1162delTAA</td>\n",
       "      <td>ENSP00000403677.1:p.Leu387_Thr388delinsSer</td>\n",
       "      <td>1884</td>\n",
       "      <td>inframe_deletion</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       #CHROM        POS   REF ALT       Consequence    Impact Symbol  \\\n",
       "58085       7  117182112  TTAA   T  inframe_deletion  MODERATE   CFTR   \n",
       "58085       7  117182112  TTAA   T  inframe_deletion  MODERATE   CFTR   \n",
       "58085       7  117182112  TTAA   T  inframe_deletion  MODERATE   CFTR   \n",
       "\n",
       "                  Gene Feature_type          Feature         Biotype  \\\n",
       "58085  ENSG00000001626   Transcript  ENST00000003084  protein_coding   \n",
       "58085  ENSG00000001626   Transcript  ENST00000426809  protein_coding   \n",
       "58085  ENSG00000001626   Transcript  ENST00000454343  protein_coding   \n",
       "\n",
       "                                     Hgvsc  \\\n",
       "58085  ENST00000003084.6:c.1160_1162delTAA   \n",
       "58085  ENST00000426809.1:c.1070_1072delTAA   \n",
       "58085  ENST00000454343.1:c.1160_1162delTAA   \n",
       "\n",
       "                                            Hgvsp Hgnc_id            effect  \n",
       "58085  ENSP00000003084.6:p.Leu387_Thr388delinsSer    1884  inframe_deletion  \n",
       "58085  ENSP00000389119.1:p.Leu357_Thr358delinsSer    1884  inframe_deletion  \n",
       "58085  ENSP00000403677.1:p.Leu387_Thr388delinsSer    1884  inframe_deletion  "
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vep_df.ix[58085]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Putative_impact</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Feature_id</th>\n",
       "      <th>Transcript_biotype</th>\n",
       "      <th>Hgvsc</th>\n",
       "      <th>Hgvsp</th>\n",
       "      <th>effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>58085</th>\n",
       "      <td>7</td>\n",
       "      <td>117182112</td>\n",
       "      <td>TTAA</td>\n",
       "      <td>T</td>\n",
       "      <td>structural_interaction_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>interaction</td>\n",
       "      <td>2PZG:B_388-B_567:ENST00000003084</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>c.1160_1162delTAA</td>\n",
       "      <td></td>\n",
       "      <td>structural_interaction_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58085</th>\n",
       "      <td>7</td>\n",
       "      <td>117182112</td>\n",
       "      <td>TTAA</td>\n",
       "      <td>T</td>\n",
       "      <td>disruptive_inframe_deletion</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000003084</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>c.1160_1162delTAA</td>\n",
       "      <td>p.Leu387_Thr388delinsSer</td>\n",
       "      <td>disruptive_inframe_deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58085</th>\n",
       "      <td>7</td>\n",
       "      <td>117182112</td>\n",
       "      <td>TTAA</td>\n",
       "      <td>T</td>\n",
       "      <td>disruptive_inframe_deletion</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000454343</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>c.1160_1162delTAA</td>\n",
       "      <td>p.Leu387_Thr388delinsSer</td>\n",
       "      <td>disruptive_inframe_deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58085</th>\n",
       "      <td>7</td>\n",
       "      <td>117182112</td>\n",
       "      <td>TTAA</td>\n",
       "      <td>T</td>\n",
       "      <td>disruptive_inframe_deletion</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000426809</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>c.1070_1072delTAA</td>\n",
       "      <td>p.Leu357_Thr358delinsSer</td>\n",
       "      <td>disruptive_inframe_deletion</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       #CHROM        POS   REF ALT                      Annotation  \\\n",
       "58085       7  117182112  TTAA   T  structural_interaction_variant   \n",
       "58085       7  117182112  TTAA   T     disruptive_inframe_deletion   \n",
       "58085       7  117182112  TTAA   T     disruptive_inframe_deletion   \n",
       "58085       7  117182112  TTAA   T     disruptive_inframe_deletion   \n",
       "\n",
       "      Putative_impact          Gene_id Feature_type  \\\n",
       "58085            HIGH  ENSG00000001626  interaction   \n",
       "58085        MODERATE  ENSG00000001626   transcript   \n",
       "58085        MODERATE  ENSG00000001626   transcript   \n",
       "58085        MODERATE  ENSG00000001626   transcript   \n",
       "\n",
       "                             Feature_id Transcript_biotype              Hgvsc  \\\n",
       "58085  2PZG:B_388-B_567:ENST00000003084     protein_coding  c.1160_1162delTAA   \n",
       "58085                   ENST00000003084     protein_coding  c.1160_1162delTAA   \n",
       "58085                   ENST00000454343     protein_coding  c.1160_1162delTAA   \n",
       "58085                   ENST00000426809     protein_coding  c.1070_1072delTAA   \n",
       "\n",
       "                          Hgvsp                          effect  \n",
       "58085                            structural_interaction_variant  \n",
       "58085  p.Leu387_Thr388delinsSer     disruptive_inframe_deletion  \n",
       "58085  p.Leu387_Thr388delinsSer     disruptive_inframe_deletion  \n",
       "58085  p.Leu357_Thr358delinsSer     disruptive_inframe_deletion  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snpeff_df.ix[58085]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "snpeff_df = snpeff_df[~snpeff_df['effect'].str.contains('structural_interaction_variant')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "collapse_map = {\n",
    "'3_prime_UTR_variant': '3_prime_UTR_variant', \n",
    "'5_prime_UTR_premature_start_codon_gain_variant': '5_prime_UTR_premature_start_codon_gain_variant',\n",
    "'5_prime_UTR_variant': '5_prime_UTR_variant',\n",
    "'coding_sequence_variant': 'coding_sequence_variant',\n",
    "'conservative_inframe_deletion': 'inframe_deletion',\n",
    "'conservative_inframe_insertion': 'inframe_insertion',\n",
    "'disruptive_inframe_deletion': 'inframe_deletion',\n",
    "'disruptive_inframe_insertion': 'inframe_insertion',\n",
    "'downstream_gene_variant': 'downstream_gene_variant',\n",
    "'exon_loss_variant': 'exon_loss_variant',\n",
    "'frameshift_variant': 'frameshift_variant',\n",
    "'incomplete_terminal_codon_variant': 'incomplete_terminal_codon_variant',\n",
    "'inframe_deletion': 'inframe_deletion',\n",
    "'inframe_insertion': 'inframe_insertion', \n",
    "'initiator_codon_variant': 'initiator_codon_variant',\n",
    "'intergenic_region': 'intergenic_region',\n",
    "'intron_variant': 'intron_variant',\n",
    "'missense_variant': 'missense_variant',\n",
    "'non_coding_transcript_exon_variant': 'non_coding_transcript_exon_variant',\n",
    "'non_coding_transcript_variant': 'non_coding_transcript_variant',\n",
    "'protein_altering_variant': 'inframe_insertion',\n",
    "'sequence_feature': 'sequence_feature',\n",
    "'splice_acceptor_variant': 'splice_acceptor_variant',\n",
    "'splice_donor_variant': 'splice_donor_variant',\n",
    "'splice_region_variant': 'splice_region_variant',\n",
    "'start_lost': 'start_lost',\n",
    "'stop_gained': 'stop_gained',\n",
    "'stop_lost': 'stop_lost', \n",
    "'stop_retained_variant': 'stop_retained_variant',\n",
    "'synonymous_variant': 'synonymous_variant',\n",
    "'upstream_gene_variant': 'upstream_gene_variant'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "vep_df['effect'] = vep_df['effect'].apply(lambda eff: collapse_map[eff])\n",
    "snpeff_df['effect'] = snpeff_df['effect'].apply(lambda eff: collapse_map[eff])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Annotations\n",
      "\n",
      "                                                     VEP    SnpEff\n",
      "3_prime_UTR_variant                               6636.0    6680.0\n",
      "5_prime_UTR_premature_start_codon_gain_variant       0.0     286.0\n",
      "5_prime_UTR_variant                               9969.0   10259.0\n",
      "coding_sequence_variant                             64.0       0.0\n",
      "downstream_gene_variant                          99177.0   99185.0\n",
      "exon_loss_variant                                    0.0       7.0\n",
      "frameshift_variant                              113706.0  113693.0\n",
      "incomplete_terminal_codon_variant                    8.0       0.0\n",
      "inframe_deletion                                 13604.0   13609.0\n",
      "inframe_insertion                                25795.0   25861.0\n",
      "initiator_codon_variant                              0.0       8.0\n",
      "intergenic_region                                    0.0    1403.0\n",
      "intron_variant                                  417918.0  416946.0\n",
      "missense_variant                                 30988.0   30988.0\n",
      "non_coding_transcript_exon_variant               29794.0   29567.0\n",
      "non_coding_transcript_variant                        0.0      42.0\n",
      "sequence_feature                                     0.0    5220.0\n",
      "splice_acceptor_variant                           2208.0    2776.0\n",
      "splice_donor_variant                              2208.0    2869.0\n",
      "splice_region_variant                            17865.0   18005.0\n",
      "start_lost                                          55.0      65.0\n",
      "stop_gained                                       4877.0    4867.0\n",
      "stop_lost                                           55.0      63.0\n",
      "stop_retained_variant                                5.0       5.0\n",
      "synonymous_variant                                9056.0    9052.0\n",
      "upstream_gene_variant                            83853.0   83809.0\n"
     ]
    }
   ],
   "source": [
    "vc_vep = vep_df.groupby(['effect']).size()\n",
    "vc_vep.name = \"VEP\"\n",
    "vc_snpeff = snpeff_df.groupby(['effect']).size()\n",
    "vc_snpeff.name = \"SnpEff\"\n",
    "vc_df = pd.DataFrame([vc_vep, vc_snpeff])\n",
    "print(\"Annotations\\n\")\n",
    "print(vc_df.transpose().fillna(0))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
